Allows sheet_for to accept an integer

tgturner opened the pull request:
Currently `sheet_for` does not allow an index for the sheet to be passed. This fix maintains compatibility by checking if a name is passed first, and then checking if the sheet's index is passed.

Fixes some bugs noticed here: https://github.com/roo-rb/roo/issues/447

chopraanmol1 commented on lib/roo/excelx.rb:
```diff
@@ -97,7 +97,7 @@ def sheets
       sheet ||= default_sheet
       validate_sheet!(sheet)
-      @sheets_by_name[sheet]
+      @sheets_by_name[sheet] || @sheets[sheet - 1]
```
I understand the reason behind using (sheet - 1) is to keep things consistent with validate_sheet!, which is a good thing.

But I think it is confusing to have two methods (sheet, sheet_for) with different starting index. i.e. sheet => 0, sheet_for => 1

What about checking the type of sheet parameter before validate_sheet! and manually adjust sheet value (sheet + 1) in case of a integer. This will keep things consistent to both validate_sheet! and sheet.

In long-term we can somehow warn in depreciation warning that we are changing starting index for sheet and sheet_for from 0 to 1, or maybe we won't need this.

tgturner commented:
@chopraanmol1 

Also, side note, using `default_sheet=` also expects 1 indexed sheets, not 0 indexed, so the `sheet` method really is the one misbehaving here

chopraanmol1 commented:
@tgturner I tried using `default_sheet=` with integer and it failed for both **Roo::OpenOffice** and **Roo::Excelx**.  Only `sheet` method worked with integer value among the limited no. of method I tested.

For Excelx most of the methods use sheet_for method for determining sheet which does not yet work with Integer value as an argument. Can you give provide a list of methods which works with 1 indexed sheet?

Note: `validate_sheet!` methods is used in 3 different methods `Base#default_sheet=`, `Excelx#sheet_for`, `OpenOffice#read_cells` and none of them (& dependent methods) handles Integer based index properly

tgturner commented:
@chopraanmol1 You are correct. I saw the comment above `default_sheet=` and assumed it worked that way. Since no places that were using `validate_sheet!` were actually accepting integers, I changed it around.

I also decided to have `default_sheet=` still set `default_sheet` with the string name to avoid any trouble having `default_sheet` be an integer might cause us down stream.

chopraanmol1 commented on test.rb:
```diff
@@ -0,0 +1,39 @@
+    # end
+    # puts Benchmark.measure{ Roo::Excelx.new('test/files/Bibelbund.xlsx').tap{|x|(2..x.last_row).each{|i| x.row(i)}} }
+  end
+end
```
Care to remove this?

tgturner replied:
Great catch! Sorry for that!

tgturner commented:
@chopraanmol1 Thanks for the feedback!

I added a couple more test cases and squashed the commits. Let me know if you were thinking that the test cases should go somewhere else or if you were looking for something different.

chopraanmol1 commented:
@tgturner new test cases looks good. It will be good to have same spec covered for sheet_for. And I think we also need to handle Roo::OpenOffice#read_cells. 

After this we can go ahead.
